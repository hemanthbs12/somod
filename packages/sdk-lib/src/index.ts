export * from "./tasks/packageJson";

export * from "./tasks/ignoreFiles";

export * from "./tasks/tsConfigBuildJson";

export * from "./tasks/init";

export * from "./tasks/build";

export * from "./tasks/link";

export * from "./tasks/dev";

export * from "./tasks/serverless";

export * from "./utils/constants";
