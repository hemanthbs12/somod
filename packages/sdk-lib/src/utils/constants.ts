export const file_packageJson = "package.json";
export const file_packageLockJson = "package-lock.json";

export const file_gitIgnore = ".gitignore";
export const file_eslintIgnore = ".eslintignore";
export const file_prettierIgnore = ".prettierignore";

export const file_dotenv = ".env";

export const file_tsConfigJson = "tsconfig.json";
export const file_tsConfigBuildJson = "tsconfig.build.json";

export const file_nextEnvDTs = "next-env.d.ts";
export const file_nextConfigJs = "next.config.js";

export const key_somod = "somod";

export const key_type = "type";

export const key_module = "module";
export const key_typings = "typings";

export const key_sideEffects = "sideEffects";

export const key_main = "main";
export const key_jsnextMain = "jsnext:main";
export const key_files = "files";

export const key_serverlessSchema = "serverlessSchema";

export const path_nodeModules = "node_modules";

export const path_build = "build";

export const path_ui = "ui";
export const path_lib = "lib";

export const path_pages = "pages";
export const path_pagesData = "pages-data";
export const path_public = "public";

export const file_index_js = "index.js";
export const file_index_dts = "index.d.ts";

export const path_git = ".git";
export const path_nextBuild = ".next";

export const file_configYaml = "config.yaml";
export const file_configJson = "config.json";

export const path_serverless = "serverless";
export const path_functions = "functions";
export const path_functionLayers = "functionLayers";
export const file_templateYaml = "template.yaml";
export const file_templateJson = "template.json";

export const path_samBuild = ".aws-sam";
export const file_samConfig = "samconfig.toml";

export const path_vercel = ".vercel";

export const cli_version_regex = /^[0-9]+\.[0-9]+\.[0-9]+$/;

export const defaultServerlessTemplateSchema =
  "https://json-schema.sodaru.com/@somod/serverless-schema/schemas/index.json";

export const defaultUiConfigSchema =
  "https://json-schema.sodaru.com/@somod/ui-config-schema/schemas/index.json";

export const defaultParametersSchema =
  "https://json-schema.sodaru.com/@somod/parameters-schema/schemas/index.json";

export const namespace_page = "UI Page";

export const namespace_public = "UI Public Asset";

export const namespace_env_config = "UI Env Config";

export const namespace_public_runtime_config = "UI Public Runtime Config";

export const namespace_server_runtime_config = "UI Server Runtime Config";

export const namespace_http_api = "Serverless Http Api";

export const namespace_export_parameter = "Serverless Export Parameter";

export const namespace_parameter = "Parameter";

export const resourceType_Function = "AWS::Serverless::Function";

export const file_parametersYaml = "parameters.yaml";

export const file_parametersJson = "parameters.json";

export const key_parameter_overrides = "parameter_overrides";
