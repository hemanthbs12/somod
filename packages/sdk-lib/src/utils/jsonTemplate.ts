import {
  intersection,
  isArray,
  isBoolean,
  isNull,
  isNumber,
  isPlainObject,
  isString
} from "lodash";
import {
  JSONType,
  JSONNode,
  JSONObjectType,
  KeywordValidator,
  KeywordProcessor
} from "@somod/types";

export const parseJson = (json: JSONType): JSONNode => {
  const navigator = (json: JSONType, parent?: JSONNode["parent"]): JSONNode => {
    if (isArray(json)) {
      const itemNodes: JSONNode[] = [];
      const thisNode: JSONNode = {
        type: "array",
        items: itemNodes,
        parent: parent
      };
      Object.freeze(thisNode);
      json.forEach((item, i) => {
        const itemNode = navigator(item, {
          key: i,
          node: thisNode
        });
        itemNodes.push(itemNode);
      });
      Object.freeze(itemNodes);
      return thisNode;
    } else if (isPlainObject(json)) {
      const propertyNodes: Record<string, JSONNode> = {};
      const thisNode: JSONNode = {
        type: "object",
        properties: propertyNodes,
        parent: parent
      };
      Object.freeze(thisNode);
      Object.keys(json).forEach(property => {
        const propertyNode = navigator(json[property], {
          key: property,
          node: thisNode
        });
        propertyNodes[property] = propertyNode;
      });
      Object.freeze(propertyNodes);
      return thisNode;
    } else if (
      isString(json) ||
      isBoolean(json) ||
      isNumber(json) ||
      isNull(json)
    ) {
      const thisNode: JSONNode = {
        type: "primitive",
        value: json,
        parent: parent
      };
      Object.freeze(thisNode);
      return thisNode;
    } else {
      throw new Error(`Unknown type ${json} ${parent}`);
    }
  };
  return navigator(json);
};

export const getPath = (jsonNode: JSONNode): (string | number)[] => {
  const path: (string | number)[] = [];
  let currentNode = jsonNode;
  while (currentNode) {
    if (currentNode.parent) {
      const key = currentNode.parent.key;
      path.unshift(key);
      currentNode = currentNode.parent.node;
    } else {
      currentNode = null;
    }
  }
  return path;
};

export const constructJson = (jsonNode: JSONNode): JSONType => {
  if (jsonNode.type == "primitive") {
    return jsonNode.value;
  } else if (jsonNode.type == "array") {
    return jsonNode.items.map(item => constructJson(item));
  } else if (jsonNode.type == "object") {
    return Object.fromEntries(
      Object.keys(jsonNode.properties).map(property => [
        property,
        constructJson(jsonNode.properties[property])
      ])
    );
  }
};

export class JSONTemplateError extends Error {
  private _node: JSONNode;
  private _error: Error;

  constructor(node: JSONNode, error: Error) {
    const message = `Error at ${getPath(node).join(".")} : ${error.message}`;
    super(message);
    this._node = node;
    this._error = error;

    // Set the prototype explicitly.
    Object.setPrototypeOf(this, new.target.prototype);
  }

  get path() {
    return getPath(this._node);
  }

  get node() {
    return this._node;
  }

  get error() {
    return this._error;
  }
}

export const validateKeywords = (
  jsonNode: JSONNode,
  keywordValidators: Record<string, KeywordValidator>
): JSONTemplateError[] => {
  const errors: JSONTemplateError[] = [];
  const keywords = Object.keys(keywordValidators);
  const navigator = (jsonNode: JSONNode) => {
    if (jsonNode.type == "array") {
      jsonNode.items.forEach(item => {
        navigator(item);
      });
    } else if (jsonNode.type == "object") {
      const properties = Object.keys(jsonNode.properties);
      properties.forEach(property => {
        navigator(jsonNode.properties[property]);
      });

      // validation
      const keywordsInThisObject = intersection(properties, keywords);
      keywordsInThisObject.forEach(keyword => {
        const keywordErrors = keywordValidators[keyword](
          keyword,
          jsonNode,
          constructJson(jsonNode.properties[keyword])
        );
        errors.push(
          ...keywordErrors.map(e => new JSONTemplateError(jsonNode, e))
        );
      });
    }
  };
  navigator(jsonNode);
  return errors;
};

export const processKeywords = (
  jsonNode: JSONNode,
  keywordProcessors: Record<string, KeywordProcessor>
): JSONType => {
  const keywords = Object.keys(keywordProcessors);
  const navigator = (jsonNode: JSONNode): JSONType => {
    if (jsonNode.type == "primitive") {
      return jsonNode.value;
    } else if (jsonNode.type == "array") {
      const jsonArray = jsonNode.items.map(item => navigator(item));
      return jsonArray;
    } else if (jsonNode.type == "object") {
      const properties = Object.keys(jsonNode.properties);
      const jsonObject = Object.fromEntries(
        properties.map(property => [
          property,
          navigator(jsonNode.properties[property])
        ])
      );

      // replacement
      const keywordsInThisObject = intersection(properties, keywords);
      const keywordReplacements = keywordsInThisObject.map(keyword => {
        try {
          const replacer = keywordProcessors[keyword](
            keyword,
            jsonNode,
            jsonObject[keyword]
          );
          return { ...replacer, keyword };
        } catch (e) {
          throw new JSONTemplateError(jsonNode, e);
        }
      });
      const keywordObjectReplacements = keywordReplacements.filter(
        r => r.type == "object"
      );
      if (keywordObjectReplacements.length > 1) {
        throw new Error(
          `Object replacement is allowed for only one keyword, Found ${keywordObjectReplacements.map(
            r => r.keyword
          )}`
        );
      }

      let resultObject =
        keywordObjectReplacements.length == 1
          ? keywordObjectReplacements[0].value
          : jsonObject;

      if (isPlainObject(resultObject)) {
        keywordReplacements.forEach(r => {
          if (r.type == "keyword") {
            delete resultObject[r.keyword];
            resultObject = { ...(resultObject as JSONObjectType), ...r.value };
          }
        });
      }
      return resultObject;
    }
  };
  return navigator(jsonNode);
};
