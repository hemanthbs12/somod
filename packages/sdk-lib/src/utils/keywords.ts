import { cloneDeep, isArray, isPlainObject, isUndefined } from "lodash";

export const getKeywordPaths = <T extends string>(
  json: unknown,
  KeywordAll: T[]
): Record<T, string[][]> => {
  const keywordPaths = Object.fromEntries(
    KeywordAll.map(keyword => [keyword, []])
  ) as Record<T, string[][]>;

  if (isPlainObject(json)) {
    KeywordAll.forEach(keyword => {
      if (!isUndefined(json[keyword])) {
        keywordPaths[keyword].push([]);
      }
    });

    Object.keys(json).forEach(key => {
      const childKeywordPaths = getKeywordPaths(json[key], KeywordAll);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(key);
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  } else if (isArray(json)) {
    json.forEach((arrayItem, index) => {
      const childKeywordPaths = getKeywordPaths(arrayItem, KeywordAll);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(index + "");
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  }
  return keywordPaths;
};

export const getKeyword = (json: unknown, path: string[]): unknown => {
  let keyword: unknown = json;

  path.forEach(pathSegment => {
    keyword = keyword[pathSegment];
  });
  return cloneDeep(keyword);
};

export const replaceKeyword = (
  json: unknown,
  path: string[],
  newValue: unknown
): void => {
  let keyword: unknown = json;
  const _path = [...path];
  const lastPathSegment = _path.pop();
  _path.forEach(pathSegment => {
    keyword = keyword[pathSegment];
  });
  keyword[lastPathSegment] = newValue;
};
