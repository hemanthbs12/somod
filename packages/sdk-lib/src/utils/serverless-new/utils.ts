import { createHash } from "crypto";

export const getNodeRuntimeVersion = () => {
  return process.env.SOMOD_SERVERLESS_NODEJS_VERSION || "16";
};

const hashModuleName = (str: string): string => {
  return createHash("sha256").update(str).digest("hex").substring(0, 8);
};

export const getSAMResourceLogicalId = (
  moduleName: string,
  somodResourceId: string
): string => {
  return "r" + hashModuleName(moduleName) + somodResourceId;
};

export const getSAMResourceName = (
  moduleName: string,
  somodResourceName: string
): unknown => {
  return {
    "Fn::Sub": [
      "somod${stackId}${moduleHash}${somodResourceName}",
      {
        stackId: {
          "Fn::Select": [2, { "Fn::Split": ["/", { Ref: "AWS::StackId" }] }]
        },
        moduleHash: hashModuleName(moduleName),
        somodResourceName: somodResourceName
      }
    ]
  };
};

export const getSAMOutputName = (exportParameterName: string) => {
  return "o" + Buffer.from(exportParameterName).toString("hex");
};

export const getParameterNameFromSAMOutputName = (samOutputName: string) => {
  return Buffer.from(samOutputName.substring(1), "hex").toString();
};
