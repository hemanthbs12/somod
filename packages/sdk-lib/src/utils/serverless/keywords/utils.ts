import { cloneDeep, isArray, isPlainObject, isUndefined } from "lodash";
import {
  KeywordAll,
  KeywordSLPDependsOn,
  KeywordSLPFunction,
  KeywordSLPRef,
  KeywordSLPRefParameter,
  KeywordSLPRefResourceName,
  SLPDependsOn,
  SLPFunction,
  SLPKeyword,
  SLPRef,
  SLPRefParameter,
  SLPRefResourceName,
  SLPTemplate
} from "../types";

const getSLPKeywordPaths = (chunk: unknown): SLPTemplate["keywordPaths"] => {
  const keywordPaths = Object.fromEntries(
    KeywordAll.map(keyword => [keyword, []])
  ) as SLPTemplate["keywordPaths"];

  if (isPlainObject(chunk)) {
    KeywordAll.forEach(keyword => {
      if (!isUndefined(chunk[keyword])) {
        keywordPaths[keyword].push([]);
      }
    });

    Object.keys(chunk).forEach(key => {
      const childKeywordPaths = getSLPKeywordPaths(chunk[key]);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(key);
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  } else if (isArray(chunk)) {
    chunk.forEach((arrayItem, index) => {
      const childKeywordPaths = getSLPKeywordPaths(arrayItem);
      KeywordAll.forEach(keyword => {
        childKeywordPaths[keyword].forEach((_keywordPaths: string[]) => {
          _keywordPaths.unshift(index + "");
          keywordPaths[keyword].push(_keywordPaths);
        });
      });
    });
  }
  return keywordPaths;
};

export const updateKeywordPathsInSLPTemplate = (
  slpTemplate: SLPTemplate
): void => {
  const slpKeywords = getSLPKeywordPaths(slpTemplate.Resources);
  slpTemplate.keywordPaths = slpKeywords;
};

export const getSLPKeyword = <T extends SLPKeyword>(
  slpTemplate: SLPTemplate,
  paths: string[]
): T => {
  let keyword: unknown = slpTemplate.Resources;
  paths.forEach(path => {
    keyword = keyword[path];
  });
  return cloneDeep(keyword) as T;
};

export const replaceSLPKeyword = (
  slpTemplate: SLPTemplate,
  paths: string[],
  newValue: unknown
): void => {
  let keyword: unknown = slpTemplate.Resources;
  const _paths = [...paths];
  const lastPath = _paths.pop();
  _paths.forEach(path => {
    keyword = keyword[path];
  });
  keyword[lastPath] = newValue;
};

/**
 * some of the slp keywords references the modulename , by default the module name is the current module of the SLP Template
 * fill the default value of modulename
 */
export const updateCurrentModuleInSLPTemplate = (
  module: string,
  slpTemplate: SLPTemplate
): void => {
  type TypeKeywordsToUpdate =
    | typeof KeywordSLPDependsOn
    | typeof KeywordSLPRef
    | typeof KeywordSLPRefParameter
    | typeof KeywordSLPRefResourceName
    | typeof KeywordSLPFunction;

  const keywordsToUpdate: TypeKeywordsToUpdate[] = [
    KeywordSLPDependsOn,
    KeywordSLPRef,
    KeywordSLPRefParameter,
    KeywordSLPRefResourceName,
    KeywordSLPFunction
  ];

  keywordsToUpdate.forEach(keyword => {
    slpTemplate.keywordPaths[keyword].forEach(keywordPaths => {
      const keywordValue = getSLPKeyword<
        | SLPDependsOn
        | SLPRef
        | SLPRefParameter
        | SLPRefResourceName
        | SLPFunction
      >(slpTemplate, keywordPaths);
      if (keyword == KeywordSLPFunction) {
        keywordValue[keyword] = {
          module,
          function: keywordValue[keyword]
        };
      } else if (keywordValue[keyword].module === undefined) {
        keywordValue[keyword].module = module;
      }
      replaceSLPKeyword(slpTemplate, keywordPaths, keywordValue);
    });
  });
};
