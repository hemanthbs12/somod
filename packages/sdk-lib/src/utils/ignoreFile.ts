import { difference, padEnd, union } from "lodash";
import { dirname, join, normalize, relative } from "path";
import { file_parametersJson, path_build, path_nodeModules } from "./constants";
import {
  readIgnoreFileStore,
  updateIgnoreFileStore,
  saveIgnoreFileStore,
  unixStylePath
} from "@solib/cli-base";
import { existsSync } from "fs";

// read from ignore file in current dir to any parent dir
// usefull in monorepo cases , .prettierignore and .eslintignore will be in parent directory at repo root, somod are at package level
const locateIgnoreFile = (dir: string, ignoreFileName: string): string => {
  const paths = unixStylePath(normalize(dir)).split("/");

  paths[0] += "/"; // correcting first path , in windows  "C:" => "C:/" , in linux "" => "/"

  for (let i = paths.length; i > 0; i--) {
    const ignoreFilePath = join(...paths.slice(0, i), ignoreFileName);
    if (existsSync(ignoreFilePath)) {
      return unixStylePath(ignoreFilePath);
    }
  }

  throw new Error(
    `${ignoreFileName} is not found in any directory in the path ${dir}`
  );
};

const relativeIgnorePaths = (
  paths: string[],
  dir: string,
  ignoreFilePath: string
): string[] => {
  const relativePath =
    "/" + unixStylePath(relative(dirname(ignoreFilePath), dir));
  return paths.map(path => {
    if (path.startsWith("/") && relativePath.length > 1) {
      path = relativePath + path;
    }
    return path;
  });
};

const defaultPaths = [path_nodeModules, path_build, `/${file_parametersJson}`];

export const validate = async (
  dir: string,
  paths: string[],
  ignoreFile: string
): Promise<void> => {
  const ignoreFilePath = locateIgnoreFile(dir, ignoreFile);
  const actualIgnorePaths = await readIgnoreFileStore(ignoreFilePath);

  const expectedIgnorePaths = relativeIgnorePaths(
    [...defaultPaths, ...paths],
    dir,
    ignoreFilePath
  );

  const _actualIgnorePaths = actualIgnorePaths.map(p => p.trim());
  const _expectedIgnorePaths = expectedIgnorePaths.map(p => p.trim());

  const missingIgnorePaths = difference(
    _expectedIgnorePaths,
    _actualIgnorePaths
  );

  if (missingIgnorePaths.length > 0) {
    throw new Error(
      `${missingIgnorePaths.join(", ")} must be in ${locateIgnoreFile(
        dir,
        ignoreFile
      )}`
    );
  }
};

export const update = async (
  dir: string,
  paths: string[],
  ignoreFile: string
): Promise<void> => {
  let ignoreFilePath: string = null;
  try {
    ignoreFilePath = locateIgnoreFile(dir, ignoreFile);
  } catch (e) {
    ignoreFilePath = null;
  }

  let actualIgnorePaths: string[] = [];

  if (ignoreFilePath == null) {
    ignoreFilePath = join(dir, ignoreFile);
  } else {
    actualIgnorePaths = await readIgnoreFileStore(ignoreFilePath);
  }

  // pad empty lines to make them unique for union operation
  actualIgnorePaths = actualIgnorePaths.map((ignorePath, i) => {
    if (ignorePath.trim() == "") {
      return padEnd("", i + 1, " ");
    }
    return ignorePath;
  });

  const pathsToBeAdded = relativeIgnorePaths(
    [...defaultPaths, ...paths],
    dir,
    ignoreFilePath
  );

  const newIgnorePaths = union(actualIgnorePaths, pathsToBeAdded).map(
    ignorePath => ignorePath.trim()
  );

  updateIgnoreFileStore(ignoreFilePath, newIgnorePaths);
};

export const save = async (dir: string, ignoreFile: string): Promise<void> => {
  let ignoreFilePath: string = null;
  try {
    ignoreFilePath = locateIgnoreFile(dir, ignoreFile);
  } catch (e) {
    ignoreFilePath = join(dir, ignoreFile);
  }

  await saveIgnoreFileStore(ignoreFilePath);
};
