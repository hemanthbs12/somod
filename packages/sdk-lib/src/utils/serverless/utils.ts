import { createHash } from "crypto";
import { getKeyword, getKeywordPaths, replaceKeyword } from "../keywords";
import { KeywordAll, SLPKeyword, SLPTemplate } from "./types";

export const updateKeywordPathsInSLPTemplate = (
  slpTemplate: SLPTemplate
): void => {
  const slpKeywords = getKeywordPaths(
    slpTemplate.original.Resources,
    KeywordAll
  );
  slpTemplate.keywordPaths = slpKeywords;
};

export const getSLPKeyword = <T extends SLPKeyword>(
  slpTemplate: SLPTemplate,
  path: string[]
): T => {
  // always return the keyword value from original
  return getKeyword(slpTemplate.original.Resources, path) as T;
};

export const replaceSLPKeyword = (
  slpTemplate: SLPTemplate,
  path: string[],
  newValue: unknown
): void => {
  replaceKeyword(slpTemplate.Resources, path, newValue);
};

const hashModuleName = (str: string): string => {
  return createHash("sha256").update(str).digest("hex").substring(0, 8);
};

export const getParameterSpaceResourceLogicalId = (
  parameterSpace: string
): string => {
  return "p" + parameterSpace;
};

export const getSAMResourceLogicalId = (
  moduleName: string,
  slpResourceId: string
): string => {
  return "r" + hashModuleName(moduleName) + slpResourceId;
};

export const getSAMResourceName = (
  moduleName: string,
  slpResourceName: string
): unknown => {
  return {
    "Fn::Sub": [
      "slp${stackId}${moduleHash}${slpResourceName}",
      {
        stackId: {
          "Fn::Select": [2, { "Fn::Split": ["/", { Ref: "AWS::StackId" }] }]
        },
        moduleHash: hashModuleName(moduleName),
        slpResourceName
      }
    ]
  };
};
