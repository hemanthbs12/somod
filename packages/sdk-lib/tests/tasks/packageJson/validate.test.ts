import { createFiles, createTempDir, deleteDir } from "../../utils";
import { validatePackageJson } from "../../../src";
import { join } from "path";

describe("Test Task validatePackageJson", () => {
  let dir: string = null;

  beforeEach(() => {
    dir = createTempDir();
  });

  afterEach(() => {
    deleteDir(dir);
  });

  const allRightPackageJson = {
    name: "@my-scope/my-module",
    version: "1.0.1",
    description: "",
    module: "build/index.js",
    typings: "build/index.d.ts",
    files: ["build"],
    sideEffects: false
  };

  const njp = {
    njp: "1.2.3"
  };

  const testData: [string, Record<string, unknown>][] = [
    ["empty", {}],
    [
      "normal",
      {
        name: "@my-scope/my-module",
        version: "1.0.1",
        description: ""
      }
    ],
    [
      "no sodaru module key",
      {
        ...allRightPackageJson
      }
    ],
    [
      "njp and somod",
      {
        ...allRightPackageJson,
        ...njp,
        somod: "1.2.3"
      }
    ],
    [
      "only njp",
      {
        ...allRightPackageJson,
        ...njp
      }
    ],
    [
      "wrong module",
      {
        ...allRightPackageJson,
        ...njp,
        module: "build/index.ts"
      }
    ],
    [
      "wrong typings",
      {
        ...allRightPackageJson,
        ...njp,
        typings: "dist/index.ts"
      }
    ],
    [
      "wrong files",
      {
        ...allRightPackageJson,
        ...njp,
        files: []
      }
    ],
    [
      "wrong sideEffects",
      {
        ...allRightPackageJson,
        ...njp,
        sideEffects: true
      }
    ],
    [
      "not-allowed keys",
      {
        ...allRightPackageJson,
        ...njp,
        main: "build/index.js",
        "jsnext:main": "build/index.js",
        type: "esm"
      }
    ],
    [
      "only somod set for njp module",
      { ...allRightPackageJson, somod: "1.2.3" }
    ]
  ];

  test.each(testData)("with %s", async (title, content) => {
    createFiles(dir, { "package.json": JSON.stringify(content) });

    let error: string;
    try {
      await validatePackageJson(dir, "njp");
    } catch (e) {
      error = e.message;
    }
    expect(
      error?.replace(
        join(dir, "package.json").split("\\").join("/"),
        "package.json"
      )
    ).toMatchSnapshot();
  });
});
