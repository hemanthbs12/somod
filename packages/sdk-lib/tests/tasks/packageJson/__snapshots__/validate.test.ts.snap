// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Task validatePackageJson with empty 1`] = `
"package.json has following errors
 must have required property 'njp'
 must have required property 'slp'
 must have required property 'somod'
 must have required property 'name'
 must have required property 'version'
 must have required property 'description'
 must have required property 'module'
 must have required property 'typings'
 must have required property 'files'
 must have required property 'sideEffects'
 exactly one of njp, slp, somod must be present"
`;

exports[`Test Task validatePackageJson with njp and somod 1`] = `
"package.json has following errors
 must have required property 'slp'
 exactly one of njp, slp, somod must be present"
`;

exports[`Test Task validatePackageJson with no sodaru module key 1`] = `
"package.json has following errors
 must have required property 'njp'
 must have required property 'slp'
 must have required property 'somod'
 exactly one of njp, slp, somod must be present"
`;

exports[`Test Task validatePackageJson with normal 1`] = `
"package.json has following errors
 must have required property 'njp'
 must have required property 'slp'
 must have required property 'somod'
 must have required property 'module'
 must have required property 'typings'
 must have required property 'files'
 must have required property 'sideEffects'
 exactly one of njp, slp, somod must be present"
`;

exports[`Test Task validatePackageJson with not-allowed keys 1`] = `
"package.json has following errors
 main must not exist
 jsnext:main must not exist
 type must not exist"
`;

exports[`Test Task validatePackageJson with only njp 1`] = `undefined`;

exports[`Test Task validatePackageJson with only somod set for njp module 1`] = `
"package.json has following errors
 somod must not exist"
`;

exports[`Test Task validatePackageJson with wrong files 1`] = `
"package.json has following errors
 files must contain build"
`;

exports[`Test Task validatePackageJson with wrong module 1`] = `
"package.json has following errors
 module must be build/index.js"
`;

exports[`Test Task validatePackageJson with wrong sideEffects 1`] = `
"package.json has following errors
 sideEffects must be false"
`;

exports[`Test Task validatePackageJson with wrong typings 1`] = `
"package.json has following errors
 typings must be build/index.d.ts"
`;
